crates/ibc/src/applications/transfer/error.rs

```diff
@@ -15,134 +15,104 @@ use crate::core::ics24_host::identifier::{ChannelId, PortId};
-
-        InvalidTracePortId
-            { pos: usize }
-            [ ValidationError ]
-            | e | { format_args!("invalid port id in trace at position: {0}", e.pos) },
-
-        InvalidTraceChannelId
-            { pos: usize }
-            [ ValidationError ]
-            | e | { format_args!("invalid channel id in trace at position: {0}", e.pos) },
-
-        InvalidTraceLength
-            { len: usize }
-            | e | { format_args!("trace length must be even but got: {0}", e.len) },
-
-        InvalidAmount
-            [ TraceError<FromDecStrErr> ]
-            | _ | { "invalid amount" },
-
-        InvalidToken
-            | _ | { "invalid token" },
-
-        Signer
-            [ SignerError ]
-            | _ | { "failed to parse signer" },
-
-        MissingDenomIbcPrefix
-            | _ | { "missing 'ibc/' prefix in denomination" },
-
-        MalformedHashDenom
-            | _ | { "hashed denom must be of the form 'ibc/{Hash}'" },
-
-        ParseHex
-            [ TraceError<EncodingError> ]
-            | _ | { "invalid hex string" },
-
-        ChannelNotUnordered
-            { order: Order }
-            | e | { format_args!("expected '{0}' channel, got '{1}'", Order::Unordered, e.order) },
-
-        InvalidVersion
-            { version: Version }
-            | e | { format_args!("expected version '{0}', got '{1}'", Version::ics20(), e.version) },
-
-        InvalidCounterpartyVersion
-            { version: Version }
-            | e | { format_args!("expected counterparty version '{0}', got '{1}'", Version::ics20(), e.version) },
-
-        CantCloseChannel
-            | _ | { "channel cannot be closed" },
-
-        PacketDataDeserialization
-            | _ | { "failed to deserialize packet data" },
-
-        AckDeserialization
-            | _ | { "failed to deserialize acknowledgement" },
-
-        ReceiveDisabled
-            | _ | { "receive is not enabled" },
-
-        SendDisabled
-            | _ | { "send is not enabled" },
-
-        ParseAccountFailure
-            | _ | { "failed to parse as AccountId" },
-
-        InvalidPort
-            { port_id: PortId, exp_port_id: PortId }
-            | e | { format_args!("invalid port: '{0}', expected '{1}'", e.port_id, e.exp_port_id) },
-
-        TraceNotFound
-            | _ | { "no trace associated with specified hash" },
-
-        DecodeRawMsg
-            [ TraceError<TendermintProtoError> ]
-            | _ | { "error decoding raw msg" },
-
-        UnknownMsgType
-            { msg_type: String }
-            | e | { format_args!("unknown msg type: {0}", e.msg_type) },
-
-        InvalidCoin
-            { coin: String }
-            | e | { format_args!("invalid coin string: {}", e.coin) },
-
-        Utf8Decode
-            [ TraceError<Utf8Error> ]
-            | _ | { "error decoding raw bytes as UTF8 string" },
-    }
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
+#[derive(Display, Debug)]
+pub enum Error {
```

Comment:
Hey thank you for starting to work on this! Could you share what changes you were planning on making? Specifically, it' not clear to me yet what the error system should look like. 

However, I was imagining we'd want to have drastically fewer error variants than this. We might even want to have all our errors just be a newtype around a `String`?

cc @hu55a1n1

Reply:
> However, I was imagining we'd want to have drastically fewer error variants than this. We might even want to have all our errors just be a newtype around a String?

I understand what you mean, for example, the displaydoc used here adds a similar comment to each variant that is the output of each display of the error variant, which is a disguised implementation of the to_string() method.

Reply:
I think we should try to restrict the error variants to those that are relevant to (or created by) this module and all other errors should be wrapped in a module variant -> e.g. `ChannelError(channel_error::Error)`. These errors can be returned as `Error::source()` in the `Error` trait impl as suggested in the [stdlib](https://doc.rust-lang.org/beta/core/error/trait.Error.html) ->
> If one module must report an error that is caused by an error from a lower-level module, it can allow accessing that error via [Error::source()](https://doc.rust-lang.org/beta/core/error/trait.Error.html#method.source). This makes it possible for the high-level module to provide its own errors while also revealing some of the implementation for debugging.